### DISCUSSION TITLE
[alpha][auth] FirebaseError: Firebase: Error (auth/already-initialized)

### DISCUSSION THREAD

**Discussion by retronav** [firebase/firebase-js-sdk#4510] [Q&A] [answered]
### Describe your environment

  * Operating System version: Windows 10
  * Browser version: 88.0.4324.182
  * Firebase SDK version: ^0.900.14
  * Firebase Product: auth

### Describe the problem

I'm setting up a basic web app using Next.js with the Firebase Alpha SDK. On the first compile, everything works fine, but as soon as subsequent hot/cold reloads occur(on code change), an error shows up like this:
Browser
![image](https://user-images.githubusercontent.com/52596591/108881148-0421e900-7629-11eb-90d7-1210c9304cd6.png)

Terminal
![image](https://user-images.githubusercontent.com/52596591/108883069-05541580-762b-11eb-9033-fa11390df889.png)

I think that this might not be an issue related with Next.js, it might occur with any module bundler/framework.
#### Relevant Code:

```tsx
import { getApp, getApps, initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";
import config from "../firebase/config.ts";
export const firebaseApp = initializeApp(config)
const auth = getAuth(firebaseApp); //This line causes an error

export default function page() {
   // Any auth logic can be here, regardless of error
   return <h1>Hello</h1>
}
```

**Comment by sam-gc** [comment, marked as answer]
This is interesting, thanks for the report. I'm not able to reproduce in vanilla JS using this code:

```js
const app = initializeApp(config);
getAuth(app);
getAuth(app);
```

But this code fails in a different way:
```js
initializeApp(config);
initializeApp(config);  // Throws FirebaseError: Firebase: Firebase App named '[DEFAULT]' already exists (app/duplicate-app).
```

Now I'm not sure how the hot reloading works and how global state is preserved / overwritten. I'll play with this a bit more if I find some time.

**Reply by hipdev** [reply]
@obnoxiousnerd Hi had the same issue in my next.js project, ^0.900.12 version fixed, I don't know also why is happening with that version, also with the last one(^0.900.15) is not working.

I am also trying this way and is still now working.

```javascript
let firebaseApp;
if (!getApps().length) {
  firebaseApp = initializeApp(firebaseConfig);
}
firebaseApp = getApp();

export const auth = getAuth(firebaseApp);
```

Hope will be solved soon

**Reply by retronav** [reply]
@samhorlbeck You'll need to add a check to ensure whether an "default" app is already registered, just like how it done in the Firebase v8 SDK
```ts
const firebaseApp = !getApps().length ? initializeApp(config) : getApp()
```
The way mentioned in @hipdev 's snippet will also work.

However, I do find an interesting phenomenon, instead of the browser version of SDK being initialized, the node version gets initialized. I guess this happens because of static generation in Next.js.
After digging a bit in the code, I see these lines:
https://github.com/firebase/firebase-js-sdk/blob/02c1761854b3037a0e2b75a5cd55d2e6c7e009f1/packages-exp/auth-exp/src/core/auth/initialize.ts#L32-L35
I think because of `_fail` being called here, the error happens.
https://github.com/firebase/firebase-js-sdk/blob/02c1761854b3037a0e2b75a5cd55d2e6c7e009f1/packages-exp/auth-exp/index.node.ts#L45-L47
Is it possible to return the current instance from here?

**Reply by retronav** [reply]
Okay guys, an update. I got it working, just the fact that it uses some internal functions:
```ts
//@ts-ignore
const authProvider = firebaseApp.container.getProvider("auth-exp");
let auth: Auth;
if (authProvider.isInitialized()) {
  auth = authProvider.getImmediate();
} else {
  auth = getAuth(firebaseApp);
}
//Do whatever with this
export {auth};
```

So maybe the `getAuth` function can be re-implemented like this(just a crude idea):
```ts
export function getAuth(app: FirebaseApp): Auth { 
   const provider = _getProvider(app, "auth-exp")
   if (provider.isInitialized()) {
    return provider.getImmediate() as AuthImpl;
  }
  else return initializeAuth(app); 
 } 
```

**Reply by Feiyang1** [reply]
That makes sense because static generation happens in Nodejs.
https://github.com/firebase/firebase-js-sdk/pull/4519 will fix the issue.

**Reply by rosnk** [reply]
in react native i did this: 

import { initializeApp, getApps, getApp } from "firebase/app";

let app;
let auth;
if (!getApps().length) {
  try {
    app = initializeApp(firebaseConfig);
    auth = initializeAuth(app, {
      persistence: getReactNativePersistence(AsyncStorage),
    });
  } catch (err) {
    console.log("error initializing");
  }
}
app = getApp();

**Reply by SrijanDas** [reply]
@rosnk , your suggestion gave me an error so I updated the code for react native (notice the else condition). Now it works perfectly fine 👌

```ruby
let app, auth;

if (!getApps().length) {
  try {
    app = initializeApp(firebaseConfig);
    auth = initializeAuth(app, {
      persistence: getReactNativePersistence(AsyncStorage),
    });
  } catch (error) {
    console.log("Error initializing app: " + error);
  }
} else {
  app = getApp();
  auth = getAuth(app);
}
`````

**Reply by gusta-of** [reply]
PERFEITO, consegui solucionar o problema, tenho uma aplicação angular com CapacitorJS e passei pelo mesmo problema ao atualizar minha versão do angular fire auth;
Basicamente alterei a configuração da injeçao do provider para 

```
 provideAuth(() => {
      const auth = getAuth();
      return auth;
      // if (Capacitor.isNativePlatform()) {
      //   return initializeAuth(getApp(), {
      //     persistence: indexedDBLocalPersistence,
      //   });
      // } else {
      //   const auth = getAuth();
      //   return auth;
      // }
    }),
```
